Corona-Warn-App/src/main/java/de/rki/coronawarnapp/ui/eventregistration/organizer/create/TraceLocationCreateViewModel.kt

```diff
@@ -0,0 +1,124 @@
+import org.joda.time.Duration
+import org.joda.time.LocalDateTime
+import java.util.Locale
+import kotlin.properties.ReadWriteProperty
+import kotlin.reflect.KProperty
+
+class TraceLocationCreateViewModel @AssistedInject constructor(
+    dispatcherProvider: DispatcherProvider,
+    @Assisted private val category: TraceLocationCategory
+) : CWAViewModel(dispatcherProvider = dispatcherProvider) {
+
+    private val mutableUiState = MutableLiveData<UIState>()
+    val uiState: LiveData<UIState>
+        get() = mutableUiState
+
+    var description: String? by UpdateDelegate()
+    var address: String? by UpdateDelegate()
+    var checkInLength: Duration? by UpdateDelegate()
+    var start: LocalDateTime? by UpdateDelegate()
+    var end: LocalDateTime? by UpdateDelegate()
+
+    init {
+        checkInLength = when (category.uiType) {
+            TraceLocationUIType.LOCATION -> {
+                Duration.standardHours(2)
+            }
+            TraceLocationUIType.EVENT -> {
+                Duration.ZERO
+            }
+        }
+    }
+
+    fun send() {
+        // TODO: This will be implemented in another PR
+    }
+
+    private fun updateState() {
+        mutableUiState.value = UIState(
+            startDate = start,
+            endDate = end,
+            lengthOfStay = checkInLength,
+            title = category.title,
+            isDateVisible = category.uiType == TraceLocationUIType.EVENT,
+            isSendEnable = when (category.uiType) {
+                TraceLocationUIType.LOCATION -> {
+                    description?.trim()?.length in 1..100 &&
+                        address?.trim()?.length in 0..100 &&
+                        (checkInLength ?: Duration.ZERO) > Duration.ZERO
+                }
+                TraceLocationUIType.EVENT -> {
+                    description?.trim()?.length in 1..100 &&
+                        address?.trim()?.length in 0..100 &&
+                        start != null &&
+                        end != null &&
+                        end?.isAfter(start) == true
+                }
+            }
+        )
+    }
+
+    data class UIState(
+        private val startDate: LocalDateTime? = null,
+        private val endDate: LocalDateTime? = null,
+        private val lengthOfStay: Duration? = null,
+        @StringRes val title: Int,
+        val isDateVisible: Boolean,
+        val isSendEnable: Boolean
+    ) {
+        fun getStartDate(locale: Locale): String? {
+            return startDate?.toString("E, ${locale.shortDatePattern()}   HH:mm", locale)
+        }
+
+        fun getEndDate(locale: Locale): String? {
+            return endDate?.toString("E, ${locale.shortDatePattern()}   HH:mm", locale)
+        }
+
+        fun getLength(resources: Resources): String? {
+            return lengthOfStay?.toReadableDuration(
+                suffix = resources.getString(R.string.tracelocation_organizer_duration_suffix)
+            )
+        }
```

Comment:
Some tests here please  for these functions ?

Reply:
- `toReadableDuration()` is already unit tested in `DurationExtensionKtTest`
- I don't think it make sense to test `LocalDateTime.toString()` as it should be already tested in joda library
- Other parts of the viewmodel are covered in `TraceLocationCreateViewModelTest`

Reply:
I mean the expected format that you display in the App. toString can  have different formats right? if you think it is not worth it , fine